File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.cpp
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& a) {
        int i=0;
        return buildBST(a,i,INT_MAX);
    }
    TreeNode *buildBST(vector<int>&a,int &i,int bound)
    {
        if(i==a.size()||a[i]>bound) return NULL;
        TreeNode* root=new TreeNode(a[i++]);
        root->left=buildBST(a,i,root->val);
        root->right=buildBST(a,i,bound);
        return root;
    }
};

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.java

class Solution {
    public TreeNode bstFromPreorder(int[] a) {
         int i=0;
        return buildBST(a,new int[]{0},Integer.MAX_VALUE);
    }
   public TreeNode buildBST(int []a,int i[],int bound)
    {
        if(i[0]==a.length||a[i[0]]>bound) return null;
        TreeNode root=new TreeNode(a[i[0]++]);
        root.left=buildBST(a,i,root.val);
        root.right=buildBST(a,i,bound);
        return root;
    }
}